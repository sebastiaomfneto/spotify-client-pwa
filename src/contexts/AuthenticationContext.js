import React, { useState, useEffect } from "react";

import { getItem, setItem } from "../services/StorageService";

export const AuthenticationContext = React.createContext();

export function AuthenticationContextProvider({ children }) {
  const initialToken = getItem("token");
  const [token, setToken] = useState(initialToken);
  const [isAuthenticated, setIsAuthenticated] = useState(!!initialToken);

  useEffect(() => {
    setItem("token", token);
    setIsAuthenticated(!!token);
  }, [token]);

  return (
    <AuthenticationContext.Provider
      value={{
        token,
        isAuthenticated,
        setToken,
      }}
    >
      {children}
    </AuthenticationContext.Provider>
  );
}
