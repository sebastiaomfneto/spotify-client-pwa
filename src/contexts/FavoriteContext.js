import React, { useEffect, useReducer } from 'react';

import * as StorageService from '../services/StorageService';
import { NotImplementedError } from '../helpers';

export const FavoriteContext = React.createContext();

export const TOGGLE_FAVORITE = 'TOGGLE_FAVORITE';

function initializer(initialState) {
  return StorageService.getItem('favorites') ?? initialState;
}

function reducer(state, action) {
  switch (action.type) {
    case TOGGLE_FAVORITE:
      const isfavorited = state.some((f) => f === action.payload);
      if (isfavorited) {
        return state.filter((favorite) => favorite !== action.payload);
      }
      return state.concat(action.payload);
    default:
      throw new NotImplementedError();
  }
}

export function FavoriteContextProvider({ initialState = [], children }) {
  const [state, dispatch] = useReducer(reducer, initialState, initializer);

  useEffect(() => {
    StorageService.setItem('favorites', state);
  }, [state]);

  return (
    <FavoriteContext.Provider
      value={{
        state,
        dispatch,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
}
