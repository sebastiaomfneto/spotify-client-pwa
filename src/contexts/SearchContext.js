import React, { useReducer } from 'react';

import { NotImplementedError } from '../helpers';

export const SearchContext = React.createContext();

export const SET_SEARCH = 'SET_SEARCH';

const initialState = '';

function reducer(_state, action) {
  switch (action.type) {
    case SET_SEARCH:
      return action.payload;
    default:
      throw new NotImplementedError();
  }
}

export function SearchContextProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <SearchContext.Provider
      value={{
        state,
        dispatch,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
}
