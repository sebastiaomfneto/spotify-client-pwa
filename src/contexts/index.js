export * from './AuthenticationContext';
export * from './FavoriteContext';
export * from './SearchContext';
