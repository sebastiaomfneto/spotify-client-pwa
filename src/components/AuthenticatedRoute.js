import React, { useContext } from 'react';
import { Redirect, Route } from 'react-router-dom';

import { AuthenticationContext } from '../contexts';

export function AuthenticatedRoute({ children, ...params }) {
  const { isAuthenticated } = useContext(AuthenticationContext);

  return (
    <Route
      {...params}
      render={() => {
        if (isAuthenticated) {
          return children;
        }

        return <Redirect to="/signin" />;
      }}
    />
  );
}
