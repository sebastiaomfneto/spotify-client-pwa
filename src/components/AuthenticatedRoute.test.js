import React from 'react';
import { render, act } from '@testing-library/react';
import { MemoryRouter, Route, Switch } from 'react-router';

import { AuthenticationContext } from '../contexts';
import { AuthenticatedRoute } from './AuthenticatedRoute';

const withProviders = (element, isAuthenticated) => (
  <MemoryRouter initialEntries={['/test']}>
    <AuthenticationContext.Provider value={{ isAuthenticated }}>
      <Switch>
        {element}
        <Route path="/signin" exact>
          <span>SIGNIN</span>
        </Route>
      </Switch>
    </AuthenticationContext.Provider>
  </MemoryRouter>
);

describe('testing AuthenticatedRoute', () => {
  test('should render children when authenticated', () => {
    let baseElement;
    act(() => {
      ({ baseElement } = render(
        withProviders(
          <AuthenticatedRoute path="/test" exact>
            <span>AUTHENTICATED</span>
          </AuthenticatedRoute>,
          true
        )
      ));
    });

    expect(baseElement).toMatchSnapshot();
  });

  test('should redirect /signin when unauthenticated', async () => {
    let baseElement;
    act(() => {
      ({ baseElement } = render(
        withProviders(
          <AuthenticatedRoute path="/test" exact>
            <span>AUTHENTICATED</span>
          </AuthenticatedRoute>,
          false
        )
      ));
    });

    expect(baseElement).toMatchSnapshot();
  });
});
