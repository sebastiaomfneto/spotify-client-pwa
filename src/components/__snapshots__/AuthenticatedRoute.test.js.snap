// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing AuthenticatedRoute should redirect /signin when unauthenticated 1`] = `
<body>
  <div>
    <span>
      SIGNIN
    </span>
  </div>
</body>
`;

exports[`testing AuthenticatedRoute should render children when authenticated 1`] = `
<body>
  <div>
    <span>
      AUTHENTICATED
    </span>
  </div>
</body>
`;
